### PR TITLE
HTBHF-1064 Added an ApplicationStartedEvent

### PR DESCRIPTION
**Pull request by YetAnotherMatt**
This refactors the Event into a new package so will break the claimant service (currently the only project using common-logging, I believe). If everyone is happy with this PR I'll prepare one for claimant-service that will bring in the new version and update the event code at the same time.

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -34,6 +34,7 @@ allprojects {
 subprojects {
     apply plugin: 'java-library'
     apply plugin: 'jacoco'
+    apply plugin: 'maven-publish'
```
Comment (YetAnotherMatt) [inline build.gradle:37]: This allows us to run `gradlew publishToMavenLocal` to publish locally for testing in other projects

File: common_logging/src/main/java/uk/gov/dhsc/htbhf/logging/event/Event.java [hunk 2]
```diff
@@ -1,20 +1,19 @@
...
 import lombok.Data;
+import uk.gov.dhsc.htbhf.logging.EventType;
 
 import java.time.LocalDateTime;
 import java.util.Collections;
 import java.util.Map;
 import java.util.TreeMap;
 
 @Data
-@Builder
```
Comment (mikeSmith08) [inline common_logging/src/main/java/uk/gov/dhsc/htbhf/logging/event/Event.java:13]: Why did you remove `@Builder`?

File: common_logging/src/test/java/uk/gov/dhsc/htbhf/logging/AbstractLoggingTest.java [hunk 3]
```diff
@@ -22,4 +22,12 @@ void assertCorrectLoggingMessage(String expectedMessage, Level expectedLevel) {
         assertThat(loggingEvent.getMessage()).isEqualTo(expectedMessage);
         assertThat(loggingEvent.getLevel().toString()).isEqualTo(expectedLevel.toString());
     }
+
+    String getSingleMessageContent(Level expectedLevel) {
```
Comment (mikeSmith08) [inline common_logging/src/test/java/uk/gov/dhsc/htbhf/logging/AbstractLoggingTest.java]: Can we combine this with the code above, there is a lot of commonality?

File: common_logging/src/test/java/uk/gov/dhsc/htbhf/logging/EventLoggerSpringConfigTest.java [hunk 4]
```diff
@@ -57,4 +61,21 @@ void shouldLogMetadataInOrderGivenDataNotInOrder() {
...
+        ApplicationStartedEvent applicationStartedEvent = ApplicationStartedEvent.builder()
+                .applicationId("myApp")
+                .applicationVersion("1.0.1")
+                .instanceIndex("1")
+                .build();
+        //When
+        eventLogger.logEvent(applicationStartedEvent);
+
+        //Then
+        String message = getSingleMessageContent(INFO);
```
Comment (mikeSmith08) [inline common_logging/src/test/java/uk/gov/dhsc/htbhf/logging/EventLoggerSpringConfigTest.java]: As per my comment above, I think that the unit tests in this class should be doing assertions in the same way, can you try to combine the two so that they are all asserting in the same way? Ok, in the tests above we are controlling the timestamp, but that aside, we should be doing assertions in the same way for each test as tey're essentially the same assertions.

File: common_logging/src/main/java/uk/gov/dhsc/htbhf/logging/event/ApplicationStartedEvent.java [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+    @Builder
+    public ApplicationStartedEvent(String applicationVersion, String applicationId, String instanceIndex) {
+        super(CommonEventType.APPLICATION_STARTED,
+                LocalDateTime.now(),
+                constructMetadata(applicationVersion, applicationId, instanceIndex));
+    }
+
+    private static Map<String, Object> constructMetadata(String applicationVersion, String applicationId, String instanceIndex) {
+        return Map.of(
+                "applicationVersion", applicationVersion,
```
Comment (mikeSmith08) [inline common_logging/src/main/java/uk/gov/dhsc/htbhf/logging/event/ApplicationStartedEvent.java]: Should these be constants, or do the same as the claimant-service and have and enum with these keys in?

File: common_logging/src/main/java/uk/gov/dhsc/htbhf/logging/event/Event.java (hunk 2 above)
Comment (YetAnotherMatt) [reply common_logging/src/main/java/uk/gov/dhsc/htbhf/logging/event/Event.java:13]: Because it prevented me from putting a builder in the subclass. I figured that the constructor only had 3 parameters, all different types, so we didn't need a builder. I'm also thinking that creating subclasses of Event seems like a nice way to go.

File: common_logging/src/main/java/uk/gov/dhsc/htbhf/logging/event/ApplicationStartedEvent.java (hunk 5 above)
Comment (YetAnotherMatt) [reply common_logging/src/main/java/uk/gov/dhsc/htbhf/logging/event/ApplicationStartedEvent.java]: I left them as strings here, as this is the only place they're mentioned (other than in the test). As long as clients use the `ApplicationStartedEvent` class there's no need to re-use the constants. I could move them into constants on the class though.

File: common_logging/src/test/java/uk/gov/dhsc/htbhf/logging/AbstractLoggingTest.java (hunk 3 above)
Comment (YetAnotherMatt) [reply common_logging/src/test/java/uk/gov/dhsc/htbhf/logging/AbstractLoggingTest.java]: refactored into `assertSingleLogMessageContainsText`

File: common_logging/src/main/java/uk/gov/dhsc/htbhf/logging/event/Event.java (hunk 2 above)
Comment (mikeSmith08) [reply common_logging/src/main/java/uk/gov/dhsc/htbhf/logging/event/Event.java:13]: Ok, if we're definitely going down the sub-class route, then should this be abstract? It does solidify the intent then.

File: common_logging/src/test/java/uk/gov/dhsc/htbhf/logging/EventLoggerSpringConfigTest.java (hunk 4 above)
Comment (mikeSmith08) [reply common_logging/src/test/java/uk/gov/dhsc/htbhf/logging/EventLoggerSpringConfigTest.java]: Would be good to still use common assertions in this unit test class.

File: common_logging/src/main/java/uk/gov/dhsc/htbhf/logging/event/Event.java (hunk 2 above)
Comment (YetAnotherMatt) [reply common_logging/src/main/java/uk/gov/dhsc/htbhf/logging/event/Event.java:13]: I hadn't intended to force the subclass to be the only way of doing this, but maybe we should talk about it and see what the general feeling is.

File: common_logging/src/test/java/uk/gov/dhsc/htbhf/logging/EventLoggerSpringConfigTest.java (hunk 4 above)
Comment (YetAnotherMatt) [reply common_logging/src/test/java/uk/gov/dhsc/htbhf/logging/EventLoggerSpringConfigTest.java]: How about `assertSingleLogMessageHasText` and `assertSingleLogMessageContainsText`? Otherwise you might have to suggest how a single assertion would work.

File: common_logging/src/main/java/uk/gov/dhsc/htbhf/logging/event/Event.java (hunk 2 above)
Comment (mikeSmith08) [reply common_logging/src/main/java/uk/gov/dhsc/htbhf/logging/event/Event.java:13]: At the moment we don't know what the general usage is as we only have two of them, but it feels like making this the equivalent of a `BaseEvent` like the BaseEntity might be good, it doesn't feel like we'd have generic events, they'll always be specific. Maybe it's too early to tell, I don't have strong feelings either way.

File: common_logging/src/test/java/uk/gov/dhsc/htbhf/logging/EventLoggerSpringConfigTest.java (hunk 4 above)
Comment (mikeSmith08) [reply common_logging/src/test/java/uk/gov/dhsc/htbhf/logging/EventLoggerSpringConfigTest.java]: That sounds a pretty good suggestion. To be honest I hadn't looked into it too deeply, it just seemed like they were doing very similar things in different ways so thought there would be a common ground they could share.

**Review by mikeSmith08 — APPROVED** [review]
